temporal-sdk/src/main/java/io/temporal/internal/sync/POJOWorkflowImplementationFactory.java

```diff
@@ -175,23 +179,30 @@ public <R> void addWorkflowImplementationFactory(
   private <T> void registerWorkflowImplementationType(
       WorkflowImplementationOptions options, Class<T> workflowImplementationClass) {
     if (DynamicWorkflow.class.isAssignableFrom(workflowImplementationClass)) {
-      addWorkflowImplementationFactory(
-          options,
-          workflowImplementationClass,
-          () -> {
+      if (dynamicWorkflowImplementationFactory != null) {
+        throw new TypeAlreadyRegisteredException(
+            "DynamicWorkflow",
+            "An implementation of DynamicWorkflow or its factory is already registered with the worker");
+      }
+      dynamicWorkflowImplementationFactory =
+          (encodedValues) -> {
             try {
-              return workflowImplementationClass.getDeclaredConstructor().newInstance();
+              try {
+                return (DynamicWorkflow)
+                    workflowImplementationClass.getDeclaredConstructor().newInstance();
+              } catch (NoSuchMethodException e) {
+                return (DynamicWorkflow)
+                    workflowImplementationClass
+                        .getDeclaredConstructor(EncodedValues.class)
+                        .newInstance(encodedValues);
+              }
```

Comment:
In Python, I am worried that making the decision automatically by inspecting the constructor signatures is too dangerous. For example, I'm worried about the behavior "flipping" without warning as a result of someone changing method resolution in a scenario with workflow class inheritance, like the example below. I think an analogous scenario exists in Java, right?

<details>

<summary>python example</summary>

```python
class Mixin:
    def __init__(self, arg: str) -> None:
        # Some unused __init__ method that someone accidentally left on this class.
        self.important_field = arg
        pass


@workflow.defn
class Base:
    def __init__(self, arg: str = "default-value") -> None:
        # Suppose someone deletes this __init__ method because they think it's
        # unnecessary. They have now switched the workflow without warning into
        # "workflow init" mode, and `important_field` will be set equal to the
        # workflow input payload.
        self.important_field = arg

    @workflow.run
    async def run(self, arg: str) -> str:
        self.important_field = "some-value"
        ...
        return self.important_field


@workflow.defn
class Workflow(Base, Mixin):
    @workflow.run
    async def run(self, arg: str) -> str:
        return self.important_field
```

</details>

Reply:
I am not  fully sure I understand this argument and if it applies to Java, but if python is also going to require an annotation then I am fine with Java requiring it to be consistent with dotnet and python